### PR TITLE
Add colours to the collection expression

### PR DESCRIPTION
**Pull request by TheLimeGlass**
### Description
Add colours to the collection expression and also clean up ancient code and rename to Sets as Items is misleading.

---
**Target Minecraft Versions:** any
**Related Issues:** https://github.com/SkriptLang/Skript/issues/1177

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprSets.java [hunk 1]
```diff
@@ -47,60 +50,60 @@
...
+@Name("Sets")
+@Description("Collection sets of items or blocks of a specific type or colours, useful for looping.")
 @Examples({"loop items of type ore and log:",
 		"	block contains loop-item",
 		"	message \"Theres at least one %loop-item% in this block\"",
 		"drop all blocks at the player # drops one of every block at the player"})
-@Since("<i>unknown</i> (before 2.1)")
-public class ExprItems extends SimpleExpression<ItemStack> {
-	
+@Since("<i>unknown</i> (before 1.4.2)")
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprSets.java:59]:
```suggestion
@Since("<i>unknown</i> (before 1.4.2), INSERT VERSION (colors)")
```

**Review by AyhamAl-Ali — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprSets.java [hunk 2]
```diff
@@ -47,60 +50,60 @@
...
+@Since("<i>unknown</i> (before 1.4.2), INSERT VERSION (colors)")
+public class ExprSets extends SimpleExpression<Object> {
+
 	static {
-		Skript.registerExpression(ExprItems.class, ItemStack.class, ExpressionType.COMBINED,
+		Skript.registerExpression(ExprSets.class, Object.class, ExpressionType.COMBINED,
 				"[(all [[of] the]|the|every)] item(s|[ ]types)", "[(all [[of] the]|the)] items of type[s] %itemtypes%",
-				"[(all [[of] the]|the|every)] block(s|[ ]types)", "[(all [[of] the]|the)] blocks of type[s] %itemtypes%");
+				"[(all [[of] the]|the|every)] block(s|[ ]types)", "[(all [[of] the]|the)] blocks of type[s] %itemtypes%",
+				"([all] [[of] the] colo[u]rs|(the|every) colo[u]r)");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSets.java:66]:
```suggestion
				"[((all [[of] the]|the) colo[u]rs|every colo[u]r)]");
```
Fixes allowing patterns like `of the colors`

File: src/main/java/ch/njol/skript/expressions/ExprSets.java [hunk 3]
```diff
@@ -118,59 +121,61 @@ public ItemStack next() {
...
 		return ItemStack.class;
 	}
 	
 	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		final Expression<ItemType> types = this.types;
-		return (blocks ? "blocks" : "items") + (types != null ? " of type" + (types.isSingle() ? "" : "s") + " " + types.toString(e, debug) : "");
+	public String toString(@Nullable Event event, boolean debug) {
+		if (event == null)
+			return "sets of data. Pattern " + pattern;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSets.java:175]: Is there a reason for this?

File: src/main/java/ch/njol/skript/expressions/ExprSets.java [hunk 4]
```diff
@@ -118,59 +121,61 @@ public ItemStack next() {
...
+		if (pattern == 4)
+			return "colours!";
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSets.java:177]:
```suggestion
			return "all colors";
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Is there a way we could make this more generic? IMO this should be possible for any enum-based ClassInfo. Stuff like spawn reasons, inventory types, etc.

File: src/main/java/ch/njol/skript/expressions/ExprSets.java (hunk 2 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprSets.java:66]: You moved the optional too far, the whole syntax is optional, I get what you mean though.

File: src/main/java/ch/njol/skript/expressions/ExprSets.java (hunk 3 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprSets.java:175]: When you have debug mode enabled it has no event and throws an error of no check with literals.

**Comment by TheLimeGlass** [comment]
> Is there a way we could make this more generic? IMO this should be possible for any enum-based ClassInfo. Stuff like spawn reasons, inventory types, etc.

Could just copy SkQuery syntax.

**Review by APickledWalrus — APPROVED** [review]
Should be good - we can always look at more generalized options in the future

File: src/main/java/ch/njol/skript/expressions/ExprSets.java [hunk 5]
```diff
@@ -47,60 +50,60 @@
 /**
  * @author Peter Güttinger
  */
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprSets.java:52]: Can be removed

**Review by AyhamAl-Ali — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprSets.java [hunk 6]
```diff
@@ -38,69 +39,68 @@
...
- * @author Peter Güttinger
- */
-@Name("Items")
-@Description("Items or blocks of a specific type, useful for looping.")
+@Name("Sets")
+@Description("Collection sets of items or blocks of a specific type or colours, useful for looping.")
 @Examples({"loop items of type ore and log:",
 		"	block contains loop-item",
 		"	message \"Theres at least one %loop-item% in this block\"",
 		"drop all blocks at the player # drops one of every block at the player"})
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprSets.java:55]:
```suggestion
@Examples({
		"loop items of type ore and log:",
				"\tblock contains loop-item",
				"\tmessage \"Theres at least one %loop-item% in this block\"",
		"drop all blocks at the player # drops one of every block at the player"
})
```

**Pull request merged**